Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm

```diff
@@ -459,6 +459,73 @@ static void replaceContentsInRange(WebCore::LocalFrame& frame, const WebCore::Si
         return;
     }
 
+    auto replacementType = m_replacementTypes.get(uuid);
+    switch (replacementType) {
+    case WebUnifiedTextReplacementType::PlainText:
+        return textReplacementSessionPerformEditActionForPlainText(*document, uuid, action);
+    case WebUnifiedTextReplacementType::RichText:
+        return textReplacementSessionPerformEditActionForRichText(*document, uuid, action);
+    }
+}
+
+void UnifiedTextReplacementController::textReplacementSessionPerformEditActionForPlainText(WebCore::Document& document, const WTF::UUID& uuid, WebKit::WebTextReplacementData::EditAction action)
+{
+    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionPerformEditActionForPlainText (%s) [action: %hhu]", uuid.toString().utf8().data(), enumToUnderlyingType(action));
+
+    RefPtr liveRange = m_contextRanges.get(uuid);
+    auto sessionRange = makeSimpleRange(liveRange);
+    if (!sessionRange) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    Vector<Ref<WebCore::Node>> sessionNodes;
+    for (Ref node : intersectingNodes(*sessionRange))
+        sessionNodes.append(node);
+
+    for (auto nodeIterator = sessionNodes.rbegin(); nodeIterator != sessionNodes.rend(); ++nodeIterator) {
+        auto node = *nodeIterator;
+        auto markers = document.markers().markersFor(node, { WebCore::DocumentMarker::Type::UnifiedTextReplacement });
+
+        for (auto markerIterator = markers.rbegin(); markerIterator != markers.rend(); ++markerIterator) {
+            auto marker = *markerIterator;
+            auto rangeToReplace = makeSimpleRange(node, *marker);
+
+            auto currentText = WebCore::plainText(rangeToReplace);
+
+            auto oldData = std::get<WebCore::DocumentMarker::UnifiedTextReplacementData>(marker->data());
+            auto offsetRange = WebCore::OffsetRange { marker->startOffset(), marker->endOffset() };
+
+            document.markers().removeMarkers(node, offsetRange, { WebCore::DocumentMarker::Type::UnifiedTextReplacement });
+
+            auto [newText, newState] = [&]() -> std::tuple<String, WebCore::DocumentMarker::UnifiedTextReplacementData::State> {
+                switch (action) {
+                case WebTextReplacementData::EditAction::Undo:
+                    return std::make_tuple(oldData.originalText, WebCore::DocumentMarker::UnifiedTextReplacementData::State::Reverted);
+
+                case WebTextReplacementData::EditAction::Redo:
+                    return std::make_tuple(oldData.originalText, WebCore::DocumentMarker::UnifiedTextReplacementData::State::Pending);
```

Comment:
(Double checking — `oldData.originalText` in the context of redo refers to the text after replacement?)

Reply:
Correct, I also realized I don't need to return a tuple here, since the first value is always the same. Will simplify that.